indexserver: update meta file if only RawConfig changes

keegancsmith opened the pull request:
This commit allows us to add/update fields like public and priority without requiring a full re-index. It uses the recently introduced ".meta" file to accomplish this.

We currently use an allowlist on fields which trigger updates. This is what we did previously, but the list has been expanded to all RawConfig fields Sourcegraph cares about. We didn't extend this functionality to all fields on Repository since some fields change between builds / it would be more work. Additionally we can't use all fields in RawConfig since some are set by the index tools themselves.

I validated the above assumptions by patching indexserver and running it on indexed-search-0 in production. See the patch and an example of the Repository diff here: https://gist.github.com/keegancsmith/0121e0cce7fb93daede78f2c0373f2cd

This commit is still missing tests. I will be adding them soon, but I have tested manually and would appreciate a review. I may merge in and follow-up with tests. Additionally there are some TODOs left in to spark some discussion. Would love to hear feedback.

Depends on https://github.com/sourcegraph/zoekt/pull/112

keegancsmith committed:
indexserver: update meta file if only RawConfig changes

This commit allows us to add/update fields like public and priority
without requiring a full re-index. It uses the recently introduced
".meta" file to accomplish this.

We currently use an allowlist on fields which trigger updates. This is
what we did previously, but the list has been expanded to all RawConfig
fields Sourcegraph cares about. We didn't extend this functionality to
all fields on Repository since some fields change between builds / it
would be more work. Additionally we can't use all fields in RawConfig
since some are set by the index tools themselves.

I validated the above assumptions by patching indexserver and running it
on indexed-search-0 in production. See the patch and an example of the
Repository diff here:
https://gist.github.com/keegancsmith/0121e0cce7fb93daede78f2c0373f2cd

This commit is still missing tests. I will be adding them soon, but I
have tested manually and would appreciate a review. I may merge in and
follow-up with tests.

Change-Id: Id5e72a4a812279b7a1b4746d29a0d92d9e019e27

stefanhengl commented on build/builder.go:
```diff
@@ -270,46 +271,80 @@ func (o *Options) shardName(n int) string {
+	// TODO(keegan) this is a sign that we need to ensure we can just check
+	// rawconfig is equal and then do a replace update. But need to validate
+	// that it is safe to do first. Will do in a follow-up.
+	for _, k := range []string{"repoid", "priority", "public", "fork", "archived"} {
```
Just looking at the RawConfig keys, "repoid" seems out of place because (1) it is already contained in type Repository and (2) it is immutable. Maybe "fork" fits that bill too? I guess RawConfig turns into "that part of metadata that doesn't require a reindex"? However, there are fields in api.Repository, like Name and URL, that don't require a reindex either. I wonder if we could make it more obvious what goes into RawConfig and what doesn't.

keegancsmith replied:
You are right. I need to rethink the merging a bit. See this diff to see what is different in practice from a pure `BuildOptions` vs what is on disk: https://gist.github.com/keegancsmith/0121e0cce7fb93daede78f2c0373f2cd

I will try to encorporate this so it isn't just about RawConfig is works on any field that should be mutable.

I think Name for now will require a reindex since we include the name in the filename. When we switch to IDs (maybe sooner rather than later?) we can have a truly immutable field for filename and can freely change name

stefanhengl commented on build/builder.go:
```diff
@@ -270,46 +271,80 @@ func (o *Options) shardName(n int) string {
+	return IndexStateEqual
+}
+
+func readMetadata(p string) (*zoekt.Repository, *zoekt.IndexMetadata, error) {
```
Maybe this should live in the zoekt package? I think it makes sense to have various convenience functions for reading metadata exported by the zoekt package.

keegancsmith replied:
I agree. I feel like this helper has been written several times in different packages.

stefanhengl commented on build/builder.go:
```diff
@@ -693,3 +728,87 @@ func (b *Builder) writeShard(fn string, ib *zoekt.IndexBuilder) (*finishedShard,
+	// TODO should this logic live in the zoekt pkg rather than the build pkg?
+	// Argument for build is its the only place we deal with writing to multiple
+	// shards. Argument for zoekt is it's the only place that understands meta
+	// implementation.
```
3rd option :-) We let it live in indexserver. It is the only place that calls it and we wouldn't have to export it.

stefanhengl commented on build/builder.go:
```diff
@@ -693,3 +728,87 @@ func (b *Builder) writeShard(fn string, ib *zoekt.IndexBuilder) (*finishedShard,
+}
+
+// jsonMarshalFileTmp will marshal v to the temporary file p + ".*.tmp" and
+// returns the file name.
```
```suggestion
// jsonMarshalFileTmp marshals v to the temporary file p + ".*.tmp" and
// returns the file name.
```

keegancsmith commented on build/builder.go:
```diff
@@ -693,3 +728,87 @@ func (b *Builder) writeShard(fn string, ib *zoekt.IndexBuilder) (*finishedShard,
+// MergeMeta updates the .meta files for the shards on disk for o.
+//
+// Note: currently we only merge the RawConfig field. If a raw config field
+// exists on disk but not in the options, we keep the on disk value.
```
I commented above, but will repeat here since this seems like the better place to discuss:

> I need to rethink the merging a bit. See this diff to see what is different in practice from a pure BuildOptions vs what is on disk: https://gist.github.com/keegancsmith/0121e0cce7fb93daede78f2c0373f2cd

That diff says why we can just make it authorative. There is important information in Repository that is computed by the builder at index time (and can't be done by Options). My thinking now is a smarter merge which is aware of what is immutable vs not.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -474,9 +479,17 @@ func (s *Server) Index(args *indexArgs) (state indexState, err error) {
 		bo := args.BuildOptions()
 		bo.SetDefaults()
-		if bo.IncrementalSkipIndexing() {
+		state := bo.IndexState()
```
I agree. The worst part is the shadowing, I will fix that. But build.IndexState is different to this indexState, so I think the similiar name is fine given namespacing. Do you have  a proposal for a better name?

IndexStateMissing is to be expected. Will always happen for a new repo. IndexStateCorrupt isn't. I will add a log line for corrupt since it is unexpected.

keegancsmith reviewed (commented):
Thanks for the review. Will try address ASAP so we can race branch cut :P

keegancsmith committed:
all: introduce ReadMetadataPath helper

This is a helper which opens the IndexFile at a path and then calls
ReadMetadata. We had this pattern several times in our codebase. This
helpers removes some boilerplate.

Change-Id: Ia4e561b2ab1742383260c1b003ceb04b877a7275

keegancsmith committed:
all: introduce Repository.MergeMutable

We use this function to decide if we can update the meta
file. Additionally it is used by IncrementalSkipIndexing to decide if
its just a meta update. This avoids us coupling logic around what
constitutes a meta update and how to do it.

We take special care to ensure some fields remain immutable, are
ignored, etc. See the code.

This was tested by running zoekt-git-index locally, mutating a
repository and observing the effects.

Change-Id: I2e1610d24c6c453374ad2f20d18a9d3ef25653c2

keegancsmith committed:
indexserver: fallback to re-indexing if meta update fails

The most likely reason this will happen is if shard 00000 is fine, but
other shards are not. Falling back to re-index is the perfect behaviour
here.

Change-Id: Ic273c4b691163429aac06a5eca798f70f170bdd6

keegancsmith commented on build/builder.go:
```diff
@@ -693,3 +728,87 @@ func (b *Builder) writeShard(fn string, ib *zoekt.IndexBuilder) (*finishedShard,
+	// implementation.
+	var todo map[string]string
+	for i := 0; ; i++ {
+		fn := o.shardName(i)
```
My bad internet lost a few comments :( I wrote agood response here, but will be more brief now. builder relies on shardName(int) since it doesn't know how many shards it will create. Additionally there may be no shards on disk (new repo). So shardNames() helper will only be used in this bit of code here. To me it seems clearer to just keep using the current signature.

However, I'll probably move this function out based on your suggestion so will try make this cleaner.

keegancsmith committed:
Update build/builder.go

Co-authored-by: Stefan Hengl <stefan@sourcegraph.com>

keegancsmith committed:
indexserver: rename state to avoid shadowing

Change-Id: Ic06b17958848a6cfe835cc9eb610bc3dbcf78512

keegancsmith committed:
indexserver: log corrupt shards

Change-Id: I3d80dddc7b1f37666748580fec858dcfb6904a92

keegancsmith committed:
build: export ShardName

Want to use this in other places.

Change-Id: I0978fb40ec29b3480dafe1d48d1ae93d06f1880a

keegancsmith committed:
build: move MergeMeta to indexserver

Only used by this command.

Change-Id: Ice4e62761e553665a8d2bc2c34963229dd9578a6

keegancsmith committed:
build: remove unused rawConfigEqual

Change-Id: Ic1e8a8158efb96b89b514e5a54ce38d3b98ad7e9

keegancsmith committed:
indexserver: initialize todo map in mergeMeta

Change-Id: Ib6f59b37bed4c9a34efac5b4465e2103e61d99a5

keegancsmith commented:
@sourcegraph/search-core I am doing extensive manual tests. I will follow-up tomorrow with comprehensive unit testing not just for this change, but indexserver (which has been lacking!). Hoping to merge in before the tests are ready.

keegancsmith committed:
indexserver: log message for meta updates

Was heard to verify difference between full updates and meta updates.

Change-Id: I887df5e3751cf2c9b6394b000e8a9225775ae41b

keegancsmith commented on build/builder.go:
```diff
@@ -270,57 +270,58 @@ func (o *Options) shardName(n int) string {
-	if !rawConfigEqual(repo.RawConfig, o.RepositoryDescription.RawConfig, "public") {
-		return false
+	// We can mutate repo since it lives in the scope of this function call.
+	if updated, err := repo.MergeMutable(&o.RepositoryDescription); err != nil {
```
via @tsenart: we never log the error, lets just return a boolean.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -474,9 +483,24 @@ func (s *Server) Index(args *indexArgs) (state indexState, err error) {
 			return indexStateNoop, nil
+
+		case build.IndexStateMeta:
+			debug.Printf("updating index.meta %s", args.String())
```
change to log.

keegancsmith committed:
review feedback

Change-Id: Ia8d0f6e46579937c452616180afac597b90669bc
